src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,176 @@
+            Exception exception = null;
+            try
+            {
+                response = await base.SendAsync(request, cancellationToken);
+                //previously Response event was only fired if request was sucessful, 
+                //this behavior does not change
+                LogHttpResponseCore(response, loggingRequestId);
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+                throw;
+            }
+            finally
+            {
+                //Stop activity if it was started and fire event to DiagnosticSource
+                if (requestActivity != null)
+                {
+                    requestActivity.SetEndTime(DateTimeStopwatch.GetTime());
+                    s_diagnosticListener.StopActivity(
+                        requestActivity, 
+                        new { Response = response, Exception = exception }
+                    );
+                }
+            }
+
+            return response;
+        }
+
+        #region private
+        private static readonly DiagnosticListener s_diagnosticListener = new DiagnosticListener(HttpHandlerLoggingStrings.DiagnosticListenerName);
+
+        private Guid LogHttpRequestCore(HttpRequestMessage request)
+        {
+            if (s_diagnosticListener.IsEnabled(HttpHandlerLoggingStrings.RequestWriteName))
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                long timestamp = Stopwatch.GetTimestamp();
+
+                s_diagnosticListener.Write(
+                    HttpHandlerLoggingStrings.RequestWriteName,
+                    new
+                    {
+                        Request = request,
+                        LoggingRequestId = loggingRequestId,
+                        Timestamp = timestamp
+                    }
+                );
+
+                return loggingRequestId;
+            }
+            return Guid.Empty;
+        }
+
+        private void LogHttpResponseCore(HttpResponseMessage response, Guid loggingRequestId)
+        {
+            // An empty loggingRequestId signifies that the request was not logged, so do
+            // not attempt to log response.
+            if (s_diagnosticListener.IsEnabled(HttpHandlerLoggingStrings.ResponseWriteName) && loggingRequestId != Guid.Empty)
+            {
+                long timestamp = Stopwatch.GetTimestamp();
+
+                s_diagnosticListener.Write(
+                    HttpHandlerLoggingStrings.ResponseWriteName,
+                    new
+                    {
+                        Response = response,
+                        LoggingRequestId = loggingRequestId,
+                        TimeStamp = timestamp
+                    }
+                );
+            }
+        }
+
+        private const string CorrelationContextHeaderName = "Correlation-Context";
+        private const string RequestIdHeaderName = "Request-Id";
```

Comment:
Maybe add a comment saying that these headers are conforming to some "standard" and nobody should just change them arbitrarilty.